contracts/governance/Governor.sol

```diff
@@ -183,9 +188,20 @@ abstract contract Governor is Context, ERC165, EIP712, IGovernor {
         uint256 proposalId,
         address account,
         uint8 support,
-        uint256 weight
+        uint256 weight,
+        bytes memory params
```

Comment:
Adding this argument is a breaking change. We'd like some way to make this backwards compatible. Think of custom counting modules that should continue to compile and work.

If we can't find a good way to make this work we can consider making the breaking change but we strongly prefer not to.

One way could be to keep the previous function signature, and add the new one with a default implementation that redirects to the old one. But then if one wants to implement a counting module using the new signature, what should they override? The new one, old one, or both, and then how? That doesn't make a lot of sense.

An alternative is to have some other contract `GovernorWithParams` where `Governor` is kept as the backwards compatible version. What I don't like is this adds quite some complexity and due to Solidity `override` syntax leaking implementation details it is probably a breaking change in another way.

Reply:
Hey @frangio, hope you had a good holiday! Coming back to this now and reviewing what you said.

> One way could be to keep the previous function signature, and add the new one with a default implementation that redirects to the old one. But then if one wants to implement a counting module using the new signature, what should they override? The new one, old one, or both, and then how? That doesn't make a lot of sense.

Yeah this seems like it'd be very unclear for a new dev looking to extend and implement new counting functionality.

> An alternative is to have some other contract GovernorWithParams where Governor is kept as the backwards compatible version. What I don't like is this adds quite some complexity and due to Solidity override syntax leaking implementation details it is probably a breaking change in another way.

Alright, I went ahead and tried this just to get a sense of what it would be like, and pushed it up to a [new branch](https://github.com/ScopeLift/openzeppelin-contracts/tree/flex-vote-extend). Here's [the diff](https://github.com/ScopeLift/openzeppelin-contracts/commit/b82a777d03ec9cb6387b755cf0ebb8ca77eb3796#). Note that I only took it as far as I needed to get it to compile. It's not cleaned up or documented, and there are failing tests.

Here are a couple of things I notice about this approach:

* I had to duplicate the logic from from the Governor's `_castVote` method in the `GovernorWithParams` new `castVoteWithParams` method, as there's no way to curry to the existing implementation without losing the parameters
* Because of this, I also had to make the proposal state `internal` rather than `private` in order to access it in `GovernorWithParams`

What do you think of this? It feels a bit clunky, but it *could* work.

> Curious if you can think of any other way to make this backwards compatible?

Hmm let me play around a bit more and see if I can come up with something

> If we can't find a good way to make this work we can consider making the breaking change but we strongly prefer not to.

I'm inclined to argue it might be worth making a breaking change here to avoid adding too much technical debt or cruft. In particular, while this is a breaking change for projects that *extend* the contracts, it's not a breaking change from an external interface standpoint. To me that feels less serious, and a reasonable tradeoff to make here. But I completely understand the importance of backwards compatibility as a principle. Not an easy call!

Reply:
Thanks for trying this! I agree it's clunky and totally agree it's adding technical debt/cruft which makes no sense for such new contracts.

We've discussed this a bit more and would be happy to make this breaking change since it's super small and contained, unlikely to affect anyone at all at this point, and because it has the potential to enable many more vote counting modules. We just need a short "How to upgrade from 4.5" section in the changelog.

Reply:
Sounds good @frangio. That sounds like the right decision to me! Thanks for consideration. I'll add that to the changelog and also continue with my implementation of #2929 to see what using this extension feels like in practice.

Reply:
Thank you! Feel free to share an early draft, I just want to make sure that this mechanism is actually good enough to implement that module.